src/GafferCycles/IECoreCyclesPreview/ShaderNetworkAlgo.cpp

```diff
@@ -773,3 +775,354 @@ IECoreScene::ShaderNetworkPtr convertLightShader( const IECoreScene::ShaderNetwo
+	{
+		// We need a 90 degree rotation about Y on the map, which we could
+		// easily add as a `tex_mapping.rotation` parameter value on
+		// `imageShader`. Except that it would get clobbered by
+		// `ShaderCache::updateShaders()` in `Renderer.cpp` - see additional
+		// comments there. So instead we insert a `mapping` shader to do that
+		// separately.
+		ShaderPtr mappingShader = new Shader( "mapping" );
+		mappingShader->parameters()[g_rotationParameter] = new V3fData( V3f( 0, -M_PI / 2.0, 0 ) );
+		const InternedString mappingHandle = network->addShader( shaderHandle.string() + "Mapping", std::move( mappingShader ) );
+		network->addConnection( ShaderNetwork::Connection( { mappingHandle, g_vectorParameter }, { imageHandle, g_vectorParameter } ) );
+		ShaderPtr geometryShader = new Shader( "geometry" );
+		const InternedString geometryHandle = network->addShader( shaderHandle.string() + "Geometry", std::move( geometryShader ) );
+		network->addConnection( ShaderNetwork::Connection( { geometryHandle, g_positionParameter }, { mappingHandle, g_vectorParameter } ) );
+	}
+}
+
+void replaceUSDShader( ShaderNetwork *network, InternedString handle, ShaderPtr &&newShader )
+{
+	// Replace original shader with the new.
+	network->setShader( handle, std::move( newShader ) );
+}
+
+} // namespace
+
+void IECoreCycles::ShaderNetworkAlgo::convertUSDShaders( ShaderNetwork *shaderNetwork )
+{
+	for( const auto &[handle, shader] : shaderNetwork->shaders() )
+	{
+		ShaderPtr newShader;
+		if( shader->getName() == "SphereLight" )
+		{
+			newShader = new Shader( "point_light", "cycles:light" );
+			transferUSDLightParameters( shaderNetwork, handle, shader.get(), newShader.get() );
+			transferUSDShapingParameters( shaderNetwork, handle, shader.get(), newShader.get() );
+			transferUSDParameter( shaderNetwork, handle, shader.get(), g_radiusParameter, newShader.get(), g_sizeParameter, 0.5f );
+			if( parameterValue( shader.get(), g_treatAsPointParameter, false ) )
+			{
+				newShader->parameters()[g_sizeParameter] = new FloatData( 0.0 );
+				newShader->parameters()[g_normalizeParameter] = new BoolData( true );
+			}
+		}
+		else if( shader->getName() == "DiskLight" )
+		{
+			newShader = new Shader( "disk_light", "cycles:light" );
+			transferUSDLightParameters( shaderNetwork, handle, shader.get(), newShader.get() );
+			transferUSDShapingParameters( shaderNetwork, handle, shader.get(), newShader.get() );
+			const float size = parameterValue( shader.get(), g_radiusParameter, 1.0f ) * 2.0f;
+			newShader->parameters()[g_widthParameter] = new FloatData( size );
+		}
+		else if( shader->getName() == "CylinderLight" )
+		{
+			// No cylinder light in Cycles, so convert to a point light of vaguely the right size.
+			newShader = new Shader( "point_light", "cycles:light" );
+			transferUSDLightParameters( shaderNetwork, handle, shader.get(), newShader.get() );
+			transferUSDShapingParameters( shaderNetwork, handle, shader.get(), newShader.get() );
```

Comment:
Ugh, indeed. I'd like to call this one out of scope too if that's OK though. It's not something I've introduced in this PR, and this will be far far easier to deal with if we introduce a specialised USDLightVisualiser.